### PR TITLE
Legal entity deactivation rework and refactoring

### PR DESCRIPTION
**Pull request by BatrakM**
- Added contracts and contract requests deactivation to legal entity deactivation
- Used taskafka for flawless deactivation process
- Refactored merge legal entity job and jobs structure in general, moved jobs to core from graphql app

### CODE REVIEW

File: apps/core/lib/core/jobs/jobs.ex [hunk 1]
```diff
@@ -0,0 +1,64 @@
+defmodule Core.Jobs do
```
Comment (PavelVesnin) [inline apps/core/lib/core/jobs/jobs.ex]: Jobs should be separated from Core. Let's create new application `Jobs` that depends on Core

File: apps/core/lib/core/employees/employee_updater.ex [hunk 2]
```diff
@@ -119,22 +124,22 @@ defmodule Core.Employees.EmployeeUpdater do
     end
   end
 
-  def set_employee_status_as_dismissed(%Employee{} = employee, headers) do
+  def set_employee_status_as_dismissed(%Employee{} = employee, user_id, skip_contracts_suspend) do
```
Comment (progsmile) [inline apps/core/lib/core/employees/employee_updater.ex]: `skip_contracts_suspend?` marking as boolean will be better

File: apps/core/lib/core/jobs/jobs.ex [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  alias TasKafka.Jobs
+
+  @merge_legal_entities_type 200
+  @legal_entity_deactivation_type 300
+
+  def type(:merge_legal_entities), do: @merge_legal_entities_type
+  def type(:legal_entity_deactivation), do: @legal_entity_deactivation_type
+
+  def prepare_mongo_filter({:status, value}, acc) do
+    Map.put(acc, "status", value |> String.to_atom() |> Job.status())
```
Comment (progsmile) [inline apps/core/lib/core/jobs/jobs.ex]: hard to read, split to 2 lines
status = value |> String.to_atom() |> Job.status()
Map.put(acc, "status", status)

File: apps/core/lib/core/jobs/legal_entity_deactivation_job.ex [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+    }
+  end
+
+  defp get_records_to_deactivate(legal_entity) do
+    %{
+      schema: "legal_entity",
+      record: legal_entity
+    }
+    |> List.wrap()
+    |> Kernel.++(get_employees_to_deactivate(legal_entity.id))
```
Comment (progsmile) [inline apps/core/lib/core/jobs/legal_entity_deactivation_job.ex]: |> Enum.concat([
   get_employees_to_deactivate(legal_entity.id),
   get_contract_requests_to_deactivate(legal_entity.id),
  get_contracts_to_deactivate(legal_entity.id)
])
|> Enum.flat_map()

File: apps/graphql/test/web/resolvers/legal_entity_deactivation_job_resolver_test.exs [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+    test "success", %{conn: conn} do
+      legal_entity = insert(:prm, :legal_entity)
+
+      job =
+        conn
+        |> post_query(@query, input_legal_entity_id(legal_entity.id))
+        |> json_response(200)
+        |> get_in(~w(data deactivateLegalEntity legalEntityDeactivationJob))
+
+      assert "PENDING" == job["status"]
```
Comment (progsmile) [inline apps/graphql/test/web/resolvers/legal_entity_deactivation_job_resolver_test.exs]: use `PENDING` status from `Job`

File: apps/core/test/unit/jobs/legal_entity_deactivation_job_test.exs [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+  alias Core.ContractRequests.RequestPack
+  alias Core.ContractRequests.CapitationContractRequest
+  alias Core.Contracts
+  alias Core.Contracts.CapitationContract
+  alias Core.Employees
+  alias Core.Employees.Employee
+  alias Core.Jobs.LegalEntityDeactivationJob
+  alias Core.LegalEntities
+  alias Core.LegalEntities.LegalEntity
+  import Mox
```
[the remark this answers is not in the document]
Comment (BatrakM) [reply apps/core/test/unit/jobs/legal_entity_deactivation_job_test.exs]: done

File: apps/graphql/test/web/resolvers/legal_entity_deactivation_job_resolver_test.exs (hunk 5 above)
Comment (BatrakM) [reply apps/graphql/test/web/resolvers/legal_entity_deactivation_job_resolver_test.exs]: done

File: apps/core/lib/core/kafka/producer.ex [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+  @moduledoc false
+
+  @deactivate_declaration_events_topic "deactivate_declaration_events"
+  @behaviour Core.Kafka.ProducerBehaviour
+
+  require Logger
+
+  def publish_deactivate_declaration_event(event) do
+    with :ok <- KafkaEx.produce(@deactivate_declaration_events_topic, 0, :erlang.term_to_binary(event)) do
+      Logger.info("Published event #{inspect(event)} to kafka", application: :kafka_ex)
```
Comment (progsmile) [inline apps/core/lib/core/kafka/producer.ex]: Messages in non-json format aren't going to be grabbed by logging system

Comment (AlexKovalevych) [reply apps/core/lib/core/kafka/producer.ex]: Correct, but logging is done wrong atm, we should use different formatter for all messages, so you don't worry about formatting inside a log message.

File: apps/core/lib/core/contracts/contracts.ex [hunk 8]
```diff
@@ -198,20 +198,29 @@ defmodule Core.Contracts do
...
-           })
-           |> PRMRepo.update(),
-         EventManager.insert_change_status(contract, contract.status, user_id) do
+         :ok <- validate_status(contract, CapitationContract.status(:verified)),
+         {:ok, contract} <- do_terminate(user_id, contract, params) do
       {:ok, contract}
     end
   end
 
+  def do_terminate(user_id, contract, params) do
```
Comment (AlexKovalevych) [inline apps/core/lib/core/contracts/contracts.ex]: why this change is required? i mean what is the key difference with previous implementation?

File: apps/core/priv/prm_repo/migrations/20190114200308_add_employees_legal_entity_id_status_is_active_index.exs [hunk 9]
```diff
@@ -0,0 +1,7 @@
+defmodule Core.PRMRepo.Migrations.AddEmployeesLegalEntityIdStatusIsActiveIndex do
+  use Ecto.Migration
+
+  def change do
+    create(index(:employees, [:legal_entity_id, :status, :is_active]))
```
Comment (AlexKovalevych) [inline apps/core/priv/prm_repo/migrations/20190114200308_add_employees_legal_entity_id_status_is_active_index.exs]: probably we should make it `if not exists` and `concurrently`

File: apps/deactivate_legal_entity_consumer/mix.exs [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+  use Mix.Project
+
+  @version "0.1.0"
+
+  def project do
+    [
+      app: :deactivate_legal_entity_consumer,
+      description: "Deactivate legal entity Kafka consumer.",
+      version: @version,
+      elixir: "~> 1.6",
```
Comment (AlexKovalevych) [inline apps/deactivate_legal_entity_consumer/mix.exs]: 1.7
